bblocks/import_tools/weo.py

```
@@ -0,0 +1,287 @@
+        raise ValueError("invalid version. Must be 1 or 2")
+
+    url = (
+        f"{BASE_URL}/en/Publications/WEO/weo-database/"
+        f"{version[0]}/{month}/download-entire-database"
+    )
+
+    response = get_response(url)
+    soup = BeautifulSoup(response.content, "html.parser")
+
+    # check is data exists
+    if soup.find("a", string="SDMX Data"):
+        return soup.find_all("a", string="SDMX Data")[0].get("href")
+
+
+@dataclass
+class WEO(ImportData):
+    """An object to extract WEO data from the IMF website using SDMX
+    To use, create an instance of the class setting the version to the desired version,
+    or leave blank to get the latest version.
+    Call the load_data method to load the data to the object. If the data is already
+    downloaded, it will load from disk, otherwise it will download the data.
+    To force an update of the data, call the update_data method.
+    To retrieve the data, call the get_data method.
+
+    Attributes:
+        version: tuple of (year, release) or "latest". Default is "latest"
+    """
+
+    version: tuple[int, int] = "latest"
+
+    def __post_init__(self):
+        """check that version is valid"""
+
+        if (self.version != "latest") and not (isinstance(self.version, tuple)):
+            raise ValueError(
+                'Invalid version. Must be a tuple of (year, release) or "latest"'
+            )
+
+    @property
+    def _path(self):
+        """Generate path based on version"""
+        return BBPaths.raw_data / f"weo_{self.version[0]}_{self.version[1]}.csv"
+
+    def _download_latest_data(self) -> None:
+        """Get data for latest release, and replace release with latest"""
+
+        href = None  # set href to none
+        # search for the latest, until data is found
+        while href is None:
+            href = get_smdx_href(self.version)
+            if href is None:
+                # roll back version
+                self.version = roll_back_version(self.version)
```

I see now about `roll_back_version`. However, this while is a bit dangerous since href could in theory keep coming back empty and roll back version could keep rolling back for a really long time.

Maybe rollback should raise an error if the year becomes too small? Or whichever way you prefer but something should break the loop.

---

yes. See above comment. It should raise an error if it tries to roll back more than once